File: leetcode/go/trees/symmetricTree(easy)/symmetricTree.go
package main

import "fmt"

/* 101. Symmetric Tree
https://leetcode.com/problems/symmetric-tree/description/

Дан корень двоичного дерева, проверьте, является ли он зеркальным отражением самого себя
(т.е. симметричен ли относительно своего центра).

Example 1:
Input: root = [1,2,2,3,4,4,3]
Output: true

Example 2:
Input: root = [1,2,2,null,3,null,3]
Output: false
*/

// Проверка симметричности бинарного дерева.
func main() {
	root1 := createTree([]interface{}{1, 2, 2, 3, 4, 4, 3})
	fmt.Println(isSymmetric(root1)) // true

	root2 := createTree([]interface{}{1, 2, 2, nil, 3, nil, 3})
	fmt.Println(isSymmetric(root2)) // false
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// isSymmetric проверяет, является ли двоичное дерево симметричным относительно своего центра.
// time: O(n), где n - количество узлов в дереве, space: O(h), где h - высота дерева.
func isSymmetric(root *TreeNode) bool {
	// Проверяем, является ли дерево пустым
	if root == nil {
		return true
	}
	// Проверяем симметричность двух поддеревьев
	return check(root.Left, root.Right)
}

// check рекурсивно проверяет симметричность двух поддеревьев.
// time: O(n), где n - количество узлов в дереве, space: O(h), где h - высота дерева.
func check(left, right *TreeNode) bool {
	// Проверяем, являются ли оба поддерева пустыми
	if left == nil || right == nil {
		return left == nil && right == nil
	}
	// Проверяем, равны ли значения корневых узлов
	if left.Val != right.Val {
		return false
	}
	return check(left.Left, right.Right) && check(left.Right, right.Left)
}

// createTree создает бинарное дерево из массива значений.
func createTree(values []interface{}) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}

	root := &TreeNode{Val: values[0].(int)} // Создаем корневой узел с первым значением в массиве
	queue := []*TreeNode{root}              // Создаем очередь для обхода дерева по уровням
	i := 1                                  // Индекс для обхода массива значений

	for len(queue) > 0 && i < len(values) {
		node := queue[0]  // Берем первый элемент из очереди
		queue = queue[1:] // Удаляем первый элемент из очереди

		// Левый потомок
		if i < len(values) && values[i] != nil {
			node.Left = &TreeNode{Val: values[i].(int)} // Создаем левый потомок и добавляем его в очередь
			queue = append(queue, node.Left)            // Добавляем левого потомка в очередь
		}
		i++

		// Правый потомок
		// Если правый потомок не равен nil, то создаем его и добавляем его в очередь
		if i < len(values) && values[i] != nil {
			node.Right = &TreeNode{Val: values[i].(int)} // Создаем правого потомка и добавляем его в очередь
			queue = append(queue, node.Right)            // Добавляем правого потомка в очередь
		}
		i++
	}

	// Возвращаем указатель на корневой узел
	return root
}

File: leetcode/java/trees/SymmetricTree/SymmetricTree.java
package leetcode.java.trees.SymmetricTree;

import java.util.LinkedList;
import java.util.Queue;

/* 101. Symmetric Tree
https://leetcode.com/problems/symmetric-tree/description/

Дан корень двоичного дерева, проверьте, является ли он зеркальным отражением самого себя
(т.е. симметричен ли относительно своего центра).

Example 1:
Input: root = [1,2,2,3,4,4,3]
Output: true

Example 2:
Input: root = [1,2,2,null,3,null,3]
Output: false
*/

public class SymmetricTree {
    // Проверка симметричности бинарного дерева
    public static void main(String[] args) {
        TreeNode root1 = createTree(new Integer[]{1, 2, 2, 3, 4, 4, 3});
        System.out.println(isSymmetric(root1)); // true

        TreeNode root2 = createTree(new Integer[]{1, 2, 2, null, 3, null, 3});
        System.out.println(isSymmetric(root2)); // false  
    }

    // Определение узла бинарного дерева
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    // isSymmetric проверяет, является ли двоичное дерево симметричным относительно своего центра.
    // time: O(n), где n - количество узлов в дереве, space: O(h), где h - высота дерева.
    private static boolean isSymmetric(TreeNode root) {
        // Проверяем, является ли дерево пустым
        if (root == null) {
            return true;
        }
        // Проверяем симметричность двух поддеревьев
        return check(root.left, root.right);
    }

    // check рекурсивно проверяет симметричность двух поддеревьев.
    private static boolean check(TreeNode left, TreeNode right) {
        // Проверяем, являются ли оба поддерева пустыми
        if (left == null || right == null) {
            return left == null && right == null;
        }
        // Проверяем, равны ли значения корневых узлов
        if (left.val != right.val) {
            return false;
        }
        // Рекурсивно проверяем левый потомок левого поддерева с правым потомком правого,
        // и правый потомок левого с левым потомком правого
        return check(left.left, right.right) && check(left.right, right.left);
    }

    // createTree создает бинарное дерево из массива значений.
    private static TreeNode createTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        TreeNode root = new TreeNode(values[0]); // Создаем корневой узел с первым значением в массиве
        Queue<TreeNode> queue = new LinkedList<>(); // Создаем очередь для обхода дерева по уровням
        queue.offer(root);
        int i = 1; // Индекс для обхода массива значений

        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll(); // Берем первый элемент из очереди

            // Левый потомок
            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]); // Создаем левого потомка
                queue.offer(node.left);              // Добавляем левого потомка в очередь
            }
            i++;

            // Правый потомок
            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]); // Создаем правого потомка
                queue.offer(node.right);              // Добавляем правого потомка в очередь
            }
            i++;
        }

        // Возвращаем ссылку на корневой узел
        return root;
    }
}
